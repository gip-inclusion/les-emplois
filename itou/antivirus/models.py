from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
