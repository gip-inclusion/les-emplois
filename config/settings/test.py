import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
