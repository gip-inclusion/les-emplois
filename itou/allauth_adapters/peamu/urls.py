from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
