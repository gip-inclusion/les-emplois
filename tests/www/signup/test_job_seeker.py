import uuid

import pytest
import respx
from allauth.account.models import EmailConfirmationHMAC
from django.conf import settings
from django.contrib import messages
from django.test import override_settings
from django.urls import reverse
from pytest_django.asserts import assertContains, assertFormError, assertMessages, assertRedirects

from itou.openid_connect.france_connect import constants as fc_constants
from itou.users.enums import UserKind
from itou.users.models import User
from itou.utils import constants as global_constants
from itou.utils.widgets import DuetDatePickerWidget
from itou.www.signup.forms import JobSeekerSituationForm
from tests.cities.factories import create_test_cities
from tests.openid_connect.france_connect.tests import FC_USERINFO, mock_oauth_dance
from tests.users.factories import DEFAULT_PASSWORD, JobSeekerFactory
from tests.utils.test import parse_response_to_soup, reload_module


class TestJobSeekerSignup:
    def setup_method(self):
        [self.city] = create_test_cities(["67"], num_per_department=1)

    def test_choose_user_kind(self, client):
        url = reverse("signup:choose_user_kind")
        response = client.get(url)
        assertContains(response, "Candidat")

        response = client.post(url, data={"kind": UserKind.JOB_SEEKER})
        assertRedirects(response, reverse("signup:job_seeker_situation"))

    def test_job_seeker_signup_situation(self, client):
        """
        Test the redirects according to the chosen situations
        """

        # Check if the form page is displayed correctly.
        url = reverse("signup:job_seeker_situation")
        response = client.get(url)
        assert response.status_code == 200

        # Check if none of the boxes are checked 'some data' needed to raise
        # form error.
        post_data = {"some": "data"}
        response = client.post(url, post_data)
        assert response.status_code == 200
        assertFormError(response.context["form"], "situation", [JobSeekerSituationForm.ERROR_NOTHING_CHECKED])

        # Check if one of eligibility criterion is checked.
        next_url = reverse("signup:job_seeker")
        for choice in JobSeekerSituationForm.ELIGIBLE_SITUATION:
            post_data = {"situation": [choice]}
            response = client.post(url, data=post_data)
            assert response.status_code == 302
            assertRedirects(response, next_url)

            post_data["situation"].append("autre")
            response = client.post(url, data=post_data)
            assert response.status_code == 302
            assertRedirects(response, next_url)

        # Check if all the eligibility criteria are checked.
        post_data = {"situation": JobSeekerSituationForm.ELIGIBLE_SITUATION}
        response = client.post(url, data=post_data)
        assert response.status_code == 302
        assertRedirects(response, next_url)

        # Check if "Autre" is the only one checked.
        post_data = {"situation": "autre"}
        response = client.post(url, data=post_data)
        assert response.status_code == 302
        next_url = reverse("signup:job_seeker_situation_not_eligible")
        assertRedirects(response, next_url)

        # Check not eligible destination page.
        url = reverse("signup:job_seeker_situation_not_eligible")
        response = client.get(url)
        assert response.status_code == 200

    def _test_job_seeker_signup_forms(self, client, nir, **extra_signup_kwargs):
        # auxiliary function tests the forms flow for JobSeeker creation with parameterized NIR
        url = reverse("signup:job_seeker")
        response = client.get(url)
        assert response.status_code == 200

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
            **extra_signup_kwargs,
        }

        response = client.post(url, data=post_data)
        assertRedirects(response, reverse("signup:job_seeker_credentials"))
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY)

        url = reverse("signup:job_seeker_credentials")
        response = client.get(url)
        assert response.status_code == 200

        post_data = {
            "password1": DEFAULT_PASSWORD,
            "password2": DEFAULT_PASSWORD,
        }
        response = client.post(url, data=post_data)
        assertRedirects(response, reverse("account_email_verification_sent"))

        # Test session cleanup
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY) is None

        user = User.objects.get(email=job_seeker_data.email)
        assert job_seeker_data.title == user.title
        assert user.has_jobseeker_profile

        return user

    def test_job_seeker_signup(self, client, snapshot, mailoutbox):
        url = reverse("signup:job_seeker")
        response = client.get(url)
        assert response.status_code == 200
        replace_in_attr = [("max", str(DuetDatePickerWidget.max_birthdate()), "2008-10-23")]
        form = parse_response_to_soup(response, selector="form.js-format-nir", replace_in_attr=replace_in_attr)
        assert str(form) == snapshot(name="job_seeker_signup_form")

        nir = "141068078200557"
        user = self._test_job_seeker_signup_forms(client, nir)
        assert user.jobseeker_profile.nir == nir

        # `username` should be a valid UUID, see `User.generate_unique_username()`.
        assert user.username == uuid.UUID(user.username, version=4).hex
        assert user.kind == UserKind.JOB_SEEKER

        # Check `EmailAddress` state.
        assert user.emailaddress_set.count() == 1
        user_email = user.emailaddress_set.first()
        assert not user_email.verified

        # Check sent email.
        assert len(mailoutbox) == 1
        email = mailoutbox[0]
        assert "Confirmez votre adresse e-mail" in email.subject
        assert "Afin de finaliser votre inscription, cliquez sur le lien suivant" in email.body
        assert email.from_email == settings.DEFAULT_FROM_EMAIL
        assert len(email.to) == 1
        assert email.to[0] == user.email

        # User cannot log in until confirmation.
        post_data = {"login": user.email, "password": DEFAULT_PASSWORD}
        url = reverse("login:job_seeker")
        response = client.post(url, data=post_data)
        assert response.status_code == 302
        assert response.url == reverse("account_email_verification_sent")

        # Confirm email + auto login.
        confirmation_token = EmailConfirmationHMAC(user_email).key
        confirm_email_url = reverse("account_confirm_email", kwargs={"key": confirmation_token})
        # User clicks on the confirm link in the email, that is a GET request.
        response = client.get(confirm_email_url)
        assertRedirects(response, reverse("welcoming_tour:index"))
        user_email = user.emailaddress_set.first()
        assert user_email.verified

        response = client.get(confirm_email_url)
        # Uses the custom template to display errors.
        assertContains(
            response,
            f"""
            <div class="alert alert-danger" role="status">
                <p class="mb-2">Ce lien de confirmation d'adresse e-mail a expir√© ou n'est pas valide.</p>
                <p class="mb-0">
                Veuillez lancer <a href="{reverse('account_email')}">une nouvelle demande de confirmation</a>.
                </p>
            </div>
            """,
            html=True,
            count=1,
        )

    def test_job_seeker_signup_temporary_nir(self, client):
        """
        For the moment, we don't handle temporary social numbers.
        Skipping NIR verification is allowed if a temporary one should be used instead.
        """

        # Temporary numbers don't have a consistent format.
        nir = "1234567895GHTUI"

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
        }

        # Temporary NIR not considered valid.
        url = reverse("signup:job_seeker")
        response = client.post(url, post_data)
        assert response.status_code == 200
        assert not response.context.get("form").is_valid()

        # Possible to submit the form without the NIR.
        user = self._test_job_seeker_signup_forms(client, nir, skip=1)

        # Temporary NIR is not stored with user information.
        assert user.jobseeker_profile.nir == ""

    def test_job_seeker_signup_temporary_nir_resubmission(self, client):
        # Temporary numbers don't have a consistent format.
        nir = "1234567895GHTUI"

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
        }

        # Temporary NIR not considered valid.
        url = reverse("signup:job_seeker")
        response = client.post(url, post_data)
        assert response.status_code == 200
        assert not response.context.get("form").is_valid()

        # Possible to submit a valid NIR and thus effect the changes
        valid_nir = "141068078200557"
        user = self._test_job_seeker_signup_forms(client, valid_nir)
        assert user.jobseeker_profile.nir == valid_nir

    def test_job_seeker_signup_temporary_nir_invalid_birthdate(self, client):
        nir = "1234567895GHTUI"

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": "Invalid birthdate",
        }

        url = reverse("signup:job_seeker")
        response = client.post(url, post_data)
        assert response.status_code == 200
        assert response.context["form"].errors == {
            "nir": ["Ce num√©ro n'est pas valide."],
            "birthdate": ["Saisissez une date valide."],
        }

        # Cannot skip the form by passing skip
        post_data["skip"] = 1
        response = client.post(url, post_data)
        assert response.status_code == 200
        assert response.context["form"].errors == {
            "birthdate": ["Saisissez une date valide."],
        }

    def test_job_seeker_signup_with_existing_email(self, client):
        alice = JobSeekerFactory(email="alice@evil.com")
        url = reverse("signup:job_seeker")
        response = client.post(
            url,
            {
                "title": "MME",
                "first_name": "Alice",
                "last_name": "Evil",
                "email": "alice@evil.com",
                "birthdate": alice.jobseeker_profile.birthdate,
                "nir": "141068078200557",
            },
        )
        assert response.status_code == 200
        assert response.context["form"].errors == {
            "email": ["Un autre utilisateur utilise d√©j√† cette adresse e-mail."]
        }

    def test_job_seeker_visit_credentials_without_session(self, client):
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY) is None

        response = client.get(reverse("signup:job_seeker_credentials"))
        assertRedirects(response, reverse("signup:job_seeker"))

    def test_job_seeker_signup_weak_password(self, client, snapshot):
        url = reverse("signup:job_seeker")
        response = client.get(url)
        assert response.status_code == 200

        job_seeker_data = JobSeekerFactory.build(for_snapshot=True)
        post_data = {
            "nir": job_seeker_data.jobseeker_profile.nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
        }

        response = client.post(url, data=post_data)
        assertRedirects(response, reverse("signup:job_seeker_credentials"))
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY)

        url = reverse("signup:job_seeker_credentials")
        response = client.get(url)
        assert response.status_code == 200

        post_data = {
            "password1": "weak_password",
            "password2": "weak_password",
        }
        response = client.post(url, data=post_data)
        assert response.status_code == 200
        assert response.context["form"].errors == {
            "password1": [
                "Ce mot de passe est trop court. Il doit contenir au minimum 14 caract√®res.",
                "Le mot de passe doit contenir au moins 3 des 4 types suivants¬†: "
                "majuscules, minuscules, chiffres, caract√®res sp√©ciaux.",
            ]
        }
        form = parse_response_to_soup(response, selector="form.js-prevent-multiple-submit")
        assert str(form) == snapshot

    @respx.mock
    @override_settings(
        FRANCE_CONNECT_BASE_URL="https://france.connect.fake",
        FRANCE_CONNECT_CLIENT_ID="IC_CLIENT_ID_123",
        FRANCE_CONNECT_CLIENT_SECRET="IC_CLIENT_SECRET_123",
    )
    @reload_module(fc_constants)
    def test_job_seeker_nir_with_france_connect(self, client):
        # NIR is set on a previous step and tested separately.
        # See self.test_job_seeker_signup
        nir = "141068078200557"
        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
        }
        response = client.post(reverse("signup:job_seeker"), data=post_data)
        assertRedirects(response, reverse("signup:job_seeker_credentials"))

        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY)

        url = reverse("signup:job_seeker_credentials")
        response = client.get(url)
        fc_url = reverse("france_connect:authorize")
        assertContains(response, fc_url)

        # New created job seeker has no title and is redirected to complete its infos
        mock_oauth_dance(client, expected_route="dashboard:edit_user_info")
        job_seeker = User.objects.get(email=FC_USERINFO["email"])
        assert nir == job_seeker.jobseeker_profile.nir
        assert job_seeker.has_jobseeker_profile

        # The session key has been removed
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY) is None

    @respx.mock
    @override_settings(
        FRANCE_CONNECT_BASE_URL="https://france.connect.fake",
        FRANCE_CONNECT_CLIENT_ID="IC_CLIENT_ID_123",
        FRANCE_CONNECT_CLIENT_SECRET="IC_CLIENT_SECRET_123",
    )
    @reload_module(fc_constants)
    def test_job_seeker_temporary_nir_with_france_connect(self, client):
        # temporary NIR is discarded on a previous step and tested separately.
        # See self.test_job_seeker_temporary_nir

        nir = ""
        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
            "skip": 1,
        }
        response = client.post(reverse("signup:job_seeker"), data=post_data)
        assertRedirects(response, reverse("signup:job_seeker_credentials"))

        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY)

        # Temporary NIR is not stored with user information.
        url = reverse("signup:job_seeker_credentials")
        response = client.get(url)
        fc_url = reverse("france_connect:authorize")
        assertContains(response, fc_url)

        # New created job seeker has no title and is redirected to complete its infos
        mock_oauth_dance(client, expected_route="dashboard:edit_user_info")
        job_seeker = User.objects.get(email=FC_USERINFO["email"])
        assert not job_seeker.jobseeker_profile.nir
        assert job_seeker.has_jobseeker_profile

    @pytest.mark.parametrize(
        "erroneous_fields,snapshot_name",
        [
            (["email"], "email_conflict"),
            (["email", "first_name", "last_name"], "email_and_name_conflict"),
            (["nir"], "nir_conflict"),
            (["email", "first_name", "last_name", "birthdate"], "missing_only_nir"),
            (["email", "nir", "first_name", "birthdate"], "missing_only_last_name"),
            (["email", "nir", "last_name", "birthdate"], "missing_only_first_name"),
            (["email", "nir", "first_name", "last_name"], "missing_only_birthdate"),
            (["nir", "first_name", "last_name", "birthdate"], "missing_only_email"),
            (["email", "nir", "first_name", "last_name", "birthdate"], "complete_match"),
            (["nir", "birthdate"], "nir_plus_mispelled_name"),
        ],
    )
    def test_job_seeker_signup_with_conflicting_fields(self, erroneous_fields, snapshot_name, client, snapshot):
        """
        Test registration error behaviour when key fields (NIR/email) conflict with existing user(s)
        A modal with detailed error information is displayed to the user
        """
        existing_user = JobSeekerFactory(for_snapshot=True)

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": job_seeker_data.jobseeker_profile.nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": str(job_seeker_data.jobseeker_profile.birthdate),
        }

        # Prepare conflict state directed by erroneous_fields parameter
        for erroneous_field in erroneous_fields:
            try:
                post_data[erroneous_field] = getattr(existing_user, erroneous_field)
            except AttributeError:
                post_data[erroneous_field] = getattr(existing_user.jobseeker_profile, erroneous_field)

        response = client.post(reverse("signup:job_seeker"), post_data)
        assert response.status_code == 200

        # Modal is rendered with expected error message according to the conflicting fields
        assertMessages(response, [messages.Message(messages.ERROR, snapshot(name=snapshot_name))])
        assert str(parse_response_to_soup(response, selector="#message-modal-1-label")) == snapshot(
            name=f"{snapshot_name}_title"
        )
        assertContains(response, reverse("login:existing_user", args=(existing_user.public_id,)))

        # NOTE: error is rendered for the case that the user ignores the modal
        if "email" in erroneous_fields:
            assert response.context["form"].errors["email"] == [
                "Un autre utilisateur utilise d√©j√† cette adresse e-mail."
            ]
        if "jobseeker_profile__nir" in erroneous_fields:
            assert response.context["form"].errors["nir"] == ["Un compte avec ce num√©ro existe d√©j√†."]

    def test_job_seeker_signup_with_conflicting_email_not_verified(self, client, snapshot):
        existing_user = JobSeekerFactory(for_snapshot=True)
        unverified_email_address = existing_user.emailaddress_set.create(email=existing_user.email, verified=False)

        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": job_seeker_data.jobseeker_profile.nir,
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": unverified_email_address.email,  # email conflict
            "birthdate": str(job_seeker_data.jobseeker_profile.birthdate),
        }

        response = client.post(reverse("signup:job_seeker"), post_data)
        assert response.status_code == 200

        # Modal is rendered with expected error message according to the conflicting fields
        assertMessages(response, [messages.Message(messages.ERROR, snapshot)])
        assert response.context["form"].errors["email"] == ["Un autre utilisateur utilise d√©j√† cette adresse e-mail."]
        assertContains(response, reverse("login:existing_user", kwargs={"user_public_id": existing_user.public_id}))

    def test_job_seeker_signup_with_conflicting_email_temporary_nir(self, client, snapshot):
        existing_user = JobSeekerFactory(for_snapshot=True)

        post_data = {
            "nir": "1234567895GHTUI",
            "title": existing_user.title,
            "first_name": existing_user.first_name,
            "last_name": existing_user.last_name,
            "email": existing_user.email,  # email conflict
            "birthdate": str(existing_user.jobseeker_profile.birthdate),
            "skip": 1,
        }

        response = client.post(reverse("signup:job_seeker"), post_data)
        assert response.status_code == 200

        assertMessages(response, [messages.Message(messages.ERROR, snapshot)])
        assert response.context["form"].errors["email"] == ["Un autre utilisateur utilise d√©j√† cette adresse e-mail."]
        assertContains(response, reverse("login:existing_user", kwargs={"user_public_id": existing_user.public_id}))

    def test_job_seeker_signup_birth_fields_conflict_temporary_nir(self, client, snapshot):
        existing_user = JobSeekerFactory(for_snapshot=True, jobseeker_profile__nir="")

        post_data = {
            "nir": "1234567895GHTUI",
            "title": existing_user.title,
            "first_name": existing_user.first_name,
            "last_name": existing_user.last_name,
            "email": "afreshemail@adomain.org",  # no email conflict
            "birthdate": str(existing_user.jobseeker_profile.birthdate),
            "skip": 1,
        }

        response = client.post(reverse("signup:job_seeker"), post_data, follow=True)

        # Non-blocking, the user can return to the signup process if it's not them
        assertRedirects(response, reverse("signup:job_seeker_credentials"))
        assertMessages(response, [messages.Message(messages.ERROR, snapshot)])
        assert str(parse_response_to_soup(response, selector="#message-modal-1-label")) == snapshot(
            name="birth_fields_conflict_title"
        )
        assertContains(response, reverse("login:existing_user", kwargs={"user_public_id": existing_user.public_id}))

    def test_job_seeker_signup_birth_fields_conflict_redefine_nir(self, client, snapshot):
        existing_user = JobSeekerFactory(for_snapshot=True, jobseeker_profile__nir="")

        post_data = {
            "nir": "141068078200557",
            "title": existing_user.title,
            "first_name": existing_user.first_name,
            "last_name": existing_user.last_name,
            "email": "afreshemail@adomain.org",  # no email conflict
            "birthdate": str(existing_user.jobseeker_profile.birthdate),
        }

        response = client.post(reverse("signup:job_seeker"), post_data, follow=True)

        # Non-blocking, the user can return to the signup process if it's not them
        assertRedirects(response, reverse("signup:job_seeker_credentials"))
        assertMessages(response, [messages.Message(messages.ERROR, snapshot)])
        assert str(parse_response_to_soup(response, selector="#message-modal-1-label")) == snapshot(
            name="birth_fields_conflict_title"
        )
        assertContains(response, reverse("login:existing_user", kwargs={"user_public_id": existing_user.public_id}))

    def test_job_seeker_signup_email_and_nir_priorities(self, client):
        """
        The NIR is normally a more reliable source of unicity than email
        When the NIR is undefined (e.g. temporary), then a matching email is more important
        """
        existing_user = JobSeekerFactory(jobseeker_profile__nir="")

        post_data = {
            "nir": "",
            "title": existing_user.title,
            "first_name": existing_user.first_name,
            "last_name": existing_user.last_name,
            "email": existing_user.email,
            "birthdate": existing_user.jobseeker_profile.birthdate,
            "skip": 1,
        }

        response = client.post(reverse("signup:job_seeker"), post_data)
        assert response.status_code == 200

        # Matching all information except the (temporary) NIR
        assert "Vous poss√©dez d√©j√† un compte" in str(
            parse_response_to_soup(response, selector="#message-modal-1-label")
        )
        assertContains(response, reverse("login:existing_user", args=(existing_user.public_id,)))

    # TODO(calum): temporary test relating to code migration, remove in the week following deployment
    def test_job_seeker_signup_temporary_redirects(self, client):
        response = client.get(reverse("signup:job_seeker_nir"))
        assertRedirects(response, reverse("signup:job_seeker_situation"))

        session = client.session
        session["job_seeker_nir"] = "141068078200557"
        session.save()

        response = client.get(reverse("signup:job_seeker"))
        assertRedirects(response, reverse("signup:job_seeker_situation"))

        session["job_seeker_nir"] = "141068078200557"
        session.save()
        response = client.get(reverse("signup:job_seeker_credentials"))
        assertRedirects(response, reverse("signup:job_seeker"))

        # Can continue with the process on the redirected page
        job_seeker_data = JobSeekerFactory.build()
        post_data = {
            "nir": "141068078200557",
            "title": job_seeker_data.title,
            "first_name": job_seeker_data.first_name,
            "last_name": job_seeker_data.last_name,
            "email": job_seeker_data.email,
            "birthdate": job_seeker_data.jobseeker_profile.birthdate,
        }

        response = client.post(response.url, data=post_data)
        assert client.session.get(global_constants.ITOU_SESSION_JOB_SEEKER_SIGNUP_KEY)
        assertRedirects(response, reverse("signup:job_seeker_credentials"))
