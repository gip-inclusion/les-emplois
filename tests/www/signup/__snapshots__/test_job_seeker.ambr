# serializer version: 1
# name: JobSeekerSignupTest.test_job_seeker_signup[job_seeker_signup_form]
  '''
  <form action="/signup/job_seeker" class="js-prevent-multiple-submit" method="post" role="form">
  
                              <input name="csrfmiddlewaretoken" type="hidden" value="NORMALIZED_CSRF_TOKEN"/>
  
                              
  
                              
  
                              <div class="form-group mb-1 form-group-required"><label class="form-label" for="id_email">Adresse e-mail</label><input autocomplete="email" class="form-control" id="id_email" maxlength="320" name="email" placeholder="adresse@email.fr" required="" type="email"/></div>
                              <div class="text-end small font-italic mb-0">
  
  
  <a data-bs-target="#no-email-modal" data-bs-toggle="modal" href="#">
      
          Pas d'adresse e-mail ?
      
  </a>
  
  <div aria-labelledby="no-email-modal-label" aria-modal="true" class="modal fade text-start" id="no-email-modal" role="dialog" tabindex="-1">
      <div class="modal-dialog modal-dialog-centered modal-lg">
          <div class="modal-content">
              <div class="modal-header">
                  <h3 class="modal-title" id="no-email-modal-label">Créer une adresse e-mail</h3>
                  <button aria-label="Fermer" class="btn-close" data-bs-dismiss="modal" type="button"></button>
              </div>
              <div class="modal-body">
                  <div class="container">
                      <div class="row">
                          <div class="col text-center">
                              <img alt="" src="/static/img/no_email.svg"/>
                          </div>
                          <div class="col">
                              <p class="pt-3">Nous vous conseillons de créer une adresse e-mail avec le service de laposte.net.</p>
                              <p>Une fois l'adresse e-mail créée, vous pourrez l'utiliser ici.</p>
                          </div>
                      </div>
                  </div>
              </div>
              <div class="modal-footer">
                  <a aria-label="Ouverture dans un nouvel onglet" class="btn btn-sm btn-primary btn-ico" data-matomo-action="clic" data-matomo-category="inscription-candidat" data-matomo-event="true" data-matomo-option="je-n-ai-pas-d-adresse-email" href="https://compte.laposte.net/inscription/index.do?srv_gestion=POLEEMPLOI" rel="noopener" target="_blank">
                      <span>Créer une adresse e-mail</span>
                      <i class="ri-external-link-line"></i>
                  </a>
              </div>
          </div>
      </div>
  </div>
  </div>
  
                              <div class="form-group mt-1 form-group-required"><label class="form-label" for="id_title">Civilité</label><select class="form-select" id="id_title" name="title" required="">
    <option value="">---------</option>
  
    <option selected="" value="M">Monsieur</option>
  
    <option value="MME">Madame</option>
  
  </select></div>
                              <div class="form-group form-group-required"><label class="form-label" for="id_last_name">Nom</label><input class="form-control" id="id_last_name" maxlength="150" name="last_name" placeholder="Durand" required="" type="text"/></div>
                              <div class="form-group form-group-required"><label class="form-label" for="id_first_name">Prénom</label><input class="form-control" id="id_first_name" maxlength="150" name="first_name" placeholder="Dominique" required="" type="text"/></div>
                              <div class="form-group form-group-required"><label class="form-label" for="id_password1">Mot de passe</label><div class="input-group">
      <input aria-describedby="id_password1_helptext" autocomplete="new-password" class="form-control" id="id_password1" name="password1" placeholder="**********" required="" type="password"/>
  
      <div class="input-group-text p-0">
          <button class="btn btn-sm btn-link btn-ico" data-it-password="toggle" type="button">
              <i aria-hidden="true" class="ri-eye-line"></i>
              <span>Afficher</span>
          </button>
      </div>
  </div><div class="form-text">Le mot de passe doit contenir au moins 3 des 4 types suivants : majuscules, minuscules, chiffres, caractères spéciaux.</div>
  </div>
                              <div class="form-group form-group-required"><label class="form-label" for="id_password2">Mot de passe (confirmation)</label><div class="input-group">
      <input autocomplete="new-password" class="form-control" id="id_password2" name="password2" placeholder="**********" required="" type="password"/>
  
      <div class="input-group-text p-0">
          <button class="btn btn-sm btn-link btn-ico" data-it-password="toggle" type="button">
              <i aria-hidden="true" class="ri-eye-line"></i>
              <span>Afficher</span>
          </button>
      </div>
  </div></div>
  
                              
                                  <div class="form-group mt-1"><label class="form-label" for="id_nir">Numéro de sécurité sociale</label><input class="form-control" disabled="" id="id_nir" name="nir" placeholder="Numéro de sécurité sociale" type="text" value="141068078200557"/></div>
                              
  
                              <div class="small pb-4 text-secondary">
                                  Vous pourrez créer et modifier des informations telles que votre adresse ou un lien vers un CV après votre inscription, à partir de votre tableau de bord.
                              </div>
  
                              
  
  
  <p>
      Pour plus d'information sur le traitement de vos données personnelles ou pour exercer vos droits, consultez
      <a aria-label="Consultez la section Protection des données (ouverture dans une nouvel onglet)" href="/legal/privacy/" rel="noopener" target="_blank">
          la section Protection des données
      </a><i class="ri-external-link-line ms-1"></i>.
  </p>
  <p>
      En cliquant sur le bouton "Inscription", vous acceptez
      <a aria-label="Acceptez les conditions générales d'utilisation (ouverture dans une nouvel onglet)" href="/legal/terms/" rel="noopener" target="_blank">
          les conditions générales d'utilisation
      </a><i class="ri-external-link-line ms-1"></i>.
  </p>
  
  
  <div class="row">
      <div class="col-12">
          <hr class="mb-3"/>
          <small class="d-inline-block mb-3">* champs obligatoires</small>
          <div class="form-row align-items-center justify-content-end gx-3">
              <div class="form-group col-12 col-lg order-3 order-lg-1">
                  
                      <a aria-label="Annuler la saisie de ce formulaire" class="btn btn-link btn-ico ps-lg-0 w-100 w-lg-auto" href="/dashboard/">
                          <i aria-hidden="true" class="ri-close-line ri-lg"></i>
                          <span>Annuler</span>
                      </a>
                  
              </div>
              
              <div class="form-group col col-lg-auto order-2 order-lg-3">
                  
                      <button aria-label="Passer à l’étape suivante" class="btn btn-block btn-primary" type="submit">
                          <span>Inscription</span>
                      </button>
                  
              </div>
          </div>
      </div>
  </div>
  
  <div aria-hidden="true" class="modal fade" id="confirm_reset_modal" tabindex="-1">
      <div class="modal-dialog modal-dialog-centered">
          <div class="modal-content">
              
                  <div class="modal-header">
                      <h3 class="modal-title">Êtes-vous sûr de vouloir annuler ?</h3>
                  </div>
                  <div class="modal-body">
                      Les informations renseignées ne seront pas enregistrées.
                      <br/>
                      Cette action est irréversible.
                  </div>
                  <div class="modal-footer">
                      <button class="btn btn-sm btn-outline-primary" data-bs-dismiss="modal" type="button">Retour</button>
                      <a class="btn btn-sm btn-danger" href="/dashboard/">Confirmer l'annulation</a>
                  </div>
              
          </div>
      </div>
  </div>
  
  
  
                          </form>
  '''
# ---
